src/org/zaproxy/zap/spider/filters/HttpPrefixFetchFilter.java

```diff
@@ -0,0 +1,320 @@
+     * @param scheme the scheme of a URI, might be {@code null}
+     * @param port the port of a URI
+     * @return {@code true} if the {@code port} is the default for the given {@code scheme}, {@code false} otherwise
+     */
+    private static boolean isDefaultHttpOrHttpsPort(String scheme, int port) {
+        if (port == DEFAULT_HTTP_PORT && isHttp(scheme)) {
+            return true;
+        }
+        if (port == DEFAULT_HTTPS_PORT && isHttps(scheme)) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Filters any URI that does not start with the defined prefix.
+     * 
+     * @return {@code FetchStatus.VALID} if the {@code uri} starts with the {@code prefix}, {@code FetchStatus.OUT_OF_SCOPE}
+     *         otherwise
+     */
+    @Override
+    public FetchStatus checkFilter(URI uri) {
+        if (uri == null) {
+            return FetchStatus.OUT_OF_SCOPE;
+        }
+
+        String otherScheme = normalisedScheme(uri.getRawScheme());
+        if (port != normalisedPort(otherScheme, uri.getPort())) {
+            return FetchStatus.OUT_OF_SCOPE;
+        }
+
+        if (!scheme.equals(otherScheme)) {
+            return FetchStatus.OUT_OF_SCOPE;
+        }
+
+        if (!hasSameHost(uri)) {
+            return FetchStatus.OUT_OF_SCOPE;
+        }
+
+        if (!startsWith(uri.getRawPath(), path)) {
+            return FetchStatus.OUT_OF_SCOPE;
+        }
+
+        return FetchStatus.VALID;
+    }
+
+    /**
+     * Tells whether or not the given {@code uri} has the same host as required by this prefix.
+     * <p>
+     * For malformed hosts it returns always {@code false}.
+     *
+     * @param uri the {@code URI} whose host will be checked
+     * @return {@code true} if the host is same, {@code false} otherwise
+     */
+    private boolean hasSameHost(URI uri) {
+        try {
+            return host.equals(normalisedHost(uri));
+        } catch (URIException e) {
+            LOGGER.warn("Failed to normalise host: " + Arrays.toString(uri.getRawHost()), e);
+        }
+        return false;
+    }
+
+    /**
+     * Tells whether or not the given {@code array} starts with the given {@code prefix}.
+     * <p>
+     * The {@code prefix} might be {@code null} in which case it's considered that the {@code array} starts with the prefix.
+     *
+     * @param array the array that will be tested if starts with the prefix, might be {@code null}
+     * @param prefix the array used as prefix, might be {@code null}
+     * @return {@code true} if the {@code array} starts with the {@code prefix}, {@code false} otherwise
+     */
+    private static boolean startsWith(char[] array, char[] prefix) {
+        if (prefix == null) {
+            return true;
+        }
+
+        if (array == null) {
+            return false;
+        }
+
+        int length = prefix.length;
+        if (array.length < length) {
+            return false;
+        }
+
+        for (int i = 0; i < length; i++) {
```

Comment:
There's no character at that index in the prefix (last character is at length -1, because of the array index starting at zero).